[lldb] Do not bump memory modificator ID when "internal" debugger memory is updated

real-mikhail opened the pull request:
This change adds a setting `target.process.track-memory-cache-changes`.
Disabling this setting prevents invalidating and updating values in `ValueObject::UpdateValueIfNeeded` when only "internal" debugger memory is updated. Writing to "internal" debugger memory happens when, for instance, expressions are evaluated by visualizers (pretty printers).
One of the examples when cache invalidation has a particularly heavy impact is visualizations of some collections: in some collections getting collection size is an expensive operation (it requires traversal of the collection).
At the same time evaluating user expression with side effects (visible to target, not only to debugger) will still bump memory ID because:

- If expression is evaluated via interpreter: it will cause write to "non-internal" memory
- If expression is JIT-compiled: then to call the function LLDB will write to "non-internal" stack memory

The downside of disabled `target.process.track-memory-cache-changes` setting is that convenience variables won't reevaluate synthetic children automatically.

Commit:
[lldb] Do not bump memory modificator ID when "internal" debugger memory is updated

This change prevents invalidating and updating values in `ValueObject::UpdateValueIfNeeded` when only "internal" debugger memory is updated.
Writing to "internal" debugger memory happens when, for instance, user expressions are evaluated by pretty printers.
For some collections getting collection size is an expensive operation (it requires traversal of the collection) and broken value caching (being invalidated after executing expressions) make things worse.
At the same time evaluating user expression with side effects (visible to target, not only to debugger) will still bump memory ID because:
1. If expression is evaluated via interpreter: it will cause write to "non-internal" memory
2. If expression is JIT-compiled: then to call the function LLDB will write to "non-internal" stack memory

jimingham commented:
When we define a variable in the expression parser, like:

(lldb) expr SomeStruct $mine = {10, 20, 30}

we allocate memory in the target (in the allocated memory cache.)  When we go to update that variable in the expression parser, we need to write a new value into the allocated memory.  Does that process still work with your change such that we see the correct new value for the variable?

Commit:
[lldb] Reformat test code and add additional test for modifying member of the struct allocated in the cache

real-mikhail commented:
I checked that and added a test for this case. Updating value works and new value can be displayed (in simple case):
```
(lldb) expr A $mine = {100, 200}
(lldb) expr $mine.a = 300
(int) $0 = 300
(lldb) expr $mine
(A) $mine = {a=300, b=200} {
  a = 300
  b = 200
}
```

But it is true that with my change memoryId will not be bumped. That means that `lldb_private::ValueObject::EvaluationPoint::NeedsUpdating` will return false. And in case if expression execution heavily modified something in convenience variable (for instance it led to changing display format or it requires reconstructing synthetic children of that variable) - then this update will not happen and stale value will be shown. I'm not sure though if this is a realistic case.

jimingham commented:
If I make an instance of a collection class (most of which are comprehended by synthetic child providers) and then call:

(lldb) expr $my_vector.append(1)
(lldb) expr $my_vector

Then if I understand correctly, the second printing will not show the elements correctly.  Our support for making std::vector objects in the expression parser is weak right now, but our plan is to make this work.  And you certainly could have the same problem with something that wrapped one of these collection classes, since its constructor would work around the problem of instantiating the template.

That doesn't seem to me like such a far-fetched example.

jimingham commented:
If this is only a problem for persistent variables, maybe we can somehow remember that we delayed an update of the memory generation and if we see that's true and we're updating a persistent variable, then clear the memory cache at that point?

real-mikhail commented:
> If I make an instance of a collection class (most of which are comprehended by synthetic child providers) and then call:
> (lldb) expr $my_vector.append(1)
> (lldb) expr $my_vector

That will still work with my changes. Because function call (`.append()`) is JIT compiled and with current implementation LLDB modifies stack memory to call such function. Approximate callstack:
```
[liblldb.dll] lldb_private::Process::WriteMemory(unsigned long long, const void *, unsigned long long, lldb_private::Status &) Process.cpp:2240
[liblldb.dll] lldb_private::Process::WriteScalarToMemory(unsigned long long, const lldb_private::Scalar &, unsigned long long, lldb_private::Status &) Process.cpp:2319
[liblldb.dll] lldb_private::Process::WritePointerToMemory(unsigned long long, unsigned long long, lldb_private::Status &) Process.cpp:2208
/* Writing return address on the stack here */ [liblldb.dll] ABIWindows_x86_64::PrepareTrivialCall(lldb_private::Thread &,unsigned long long,unsigned long long,unsigned long long,ArrayRef<unsigned __int64>) ABIWindows_x86_64.cpp:1163
[liblldb.dll] lldb_private::ThreadPlanCallFunction::ThreadPlanCallFunction(lldb_private::Thread &,const lldb_private::Address &,const lldb_private::CompilerType &,ArrayRef<unsigned __int64>,const lldb_private::EvaluateExpressionOptions &) ThreadPlanCallFunction.cpp:143
[liblldb.dll] lldb_private::ThreadPlanCallUserExpression::ThreadPlanCallUserExpression(lldb_private::Thread &,lldb_private::Address &,ArrayRef<unsigned __int64>,const lldb_private::EvaluateExpressionOptions &,std::shared_ptr<lldb_private::UserExpression> &) ThreadPlanCallUserExpression.cpp:38
[liblldb.dll] lldb_private::LLVMUserExpression::DoExecute(lldb_private::DiagnosticManager &, lldb_private::ExecutionContext &, lldb_private::ValueObject *, const lldb_private::EvaluateExpressionOptions &, std::shared_ptr<…> &, std::shared_ptr<…> &) LLVMUserExpression.cpp:158
[liblldb.dll] lldb_private::UserExpression::Execute(lldb_private::DiagnosticManager &, lldb_private::ExecutionContext &, lldb_private::ValueObject *, const lldb_private::EvaluateExpressionOptions &, std::shared_ptr<…> &, std::shared_ptr<…> &) UserExpression.cpp:550
[liblldb.dll] lldb_private::ExpressionCache::Execute(lldb_private::ExecutionContext &, lldb_private::ValueObject *, shared_ptr<…>, const lldb_private::EvaluateExpressionOptions &, lldb_private::SharingPtr<…> &) ExpressionCache.cpp:320
[liblldb.dll] lldb_private::ExpressionCache::EvaluateWithCache(lldb_private::ExecutionContext &, const lldb_private::EvaluateExpressionOptions &, StringRef, lldb_private::SharingPtr<…> &, lldb_private::ValueObject *, const std::shared_ptr<…> &, lldb_private::ExpressionCache::ExpressionKey &&) ExpressionCache.cpp:412
[liblldb.dll] lldb_private::ExpressionCache::Evaluate(lldb_private::ExecutionContext &, const lldb_private::EvaluateExpressionOptions &, StringRef, lldb_private::SharingPtr<…> &, std::string *, lldb_private::ValueObject *, shared_ptr<…>) ExpressionCache.cpp:208
[liblldb.dll] lldb_private::Target::EvaluateExpression(StringRef, lldb_private::ExecutionContextScope *, lldb_private::SharingPtr<…> &, const lldb_private::EvaluateExpressionOptions &, std::string *, lldb_private::ValueObject *) Target.cpp:2471
```
(this is from older LLDB version so line numbers might be wrong)

The case where it won't work is something like variant class. For instance, if there is a struct with 2 fields and first field decodes how to display the second field, and there is a pretty printer which creates a synthetic child for second field, and in `expr` user directly assigns some value to one of the fields (and not calls a property to set it) - then LLDB won't show correct values.


> If this is only a problem for persistent variables, maybe we can somehow remember that we delayed an update of the memory generation and if we see that's true and we're updating a persistent variable, then clear the memory cache at that point?

In general case persistent variables can holds pointers to non-persistent ones and vice versa. I guess we can have two `MemoryId` variables (second one for memory cache) and if we can easily understand whether `ValueObject` is a persistent variable or not, then we can check whether variable needs updating by checking corresponding `MemoryId`. I think something like that might work.

real-mikhail commented:
> I think something like that might work.

Clarification: that will work only for cases if there are `VariableObject` for each variable and they are checked. If, for instance, non-persistent variable holds reference to persistent variable and only `MemoryCacheId` is bumped then (pretty printer of) non-persistent variable will not know that it needs updating.

jimingham commented:
> > I think something like that might work.
> 
> Clarification: that will work only for cases if there are `VariableObject` for each variable and they are checked. If, for instance, non-persistent variable holds reference to persistent variable and only `MemoryCacheId` is bumped then (pretty printer of) non-persistent variable will not know that it needs updating.

Making an object an object in the expression parser so that you can assign it to a member of some extant object is a useful way to test "what happens if an object made this way is put here".  That's still a pretty advanced use, so if for instance you had to set a "target.process.expressions-flush-memory-cache" setting for the times we get this wrong.  It seems like you are also solving a performance problem that most people don't have - we try hard not to use expressions in our data formatters.  

So having a way to indicate you need the faster vrs. the more bullet-proof behavior might be an acceptable solution.  You might even argue that the setting should be "bullet-proof behavior" by default.  If you are shipping a data formatter that benefits from this setting, you can also set the setting in the module that installs the data formatters...

labath commented:
> So having a way to indicate you need the faster vrs. the more bullet-proof behavior might be an acceptable solution. You might even argue that the setting should be "bullet-proof behavior" by default. If you are shipping a data formatter that benefits from this setting, you can also set the setting in the module that installs the data formatters...

Or make it an expression flag that can be set (in SBExpressionOptions) on expressions that you know are safe.

(But generally yes, it's better (and faster) to avoid expression evaluation altogether. I don't know how many data formatters are you dealing with, but I would recommend changing them to not do that.)

jimingham commented:
> On Mar 20, 2025, at 1:47 AM, Pavel Labath ***@***.***> wrote:
> 
> 
> labath
>  left a comment 
> (llvm/llvm-project#129092)
> So having a way to indicate you need the faster vrs. the more bullet-proof behavior might be an acceptable solution. You might even argue that the setting should be "bullet-proof behavior" by default. If you are shipping a data formatter that benefits from this setting, you can also set the setting in the module that installs the data formatters...
> 
> Or make it an expression flag that can be set (in SBExpressionOptions) on expressions that you know are safe.
> 
>  <https://github.com/llvm/llvm-project/pull/129092#issuecomment-2739612203> <https://github.com/notifications/unsubscribe-auth/ADUPVW3GXGYL53DSBSA5ZET2VJ6CXAVCNFSM6AAAAABYAO4MDSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMZZGYYTEMRQGM>
> 
> labath
>  left a comment 
> (llvm/llvm-project#129092)
>  <https://github.com/llvm/llvm-project/pull/129092#issuecomment-2739612203>
> So having a way to indicate you need the faster vrs. the more bullet-proof behavior might be an acceptable solution. You might even argue that the setting should be "bullet-proof behavior" by default. If you are shipping a data formatter that benefits from this setting, you can also set the setting in the module that installs the data formatters...
> 
> Or make it an expression flag that can be set (in SBExpressionOptions) on expressions that you know are safe.
> 
Yes, that's certainly a better way to do it, since then we can make the default in SBExpressionOptions be "bullet-proof" and fast behavior can be dialed up in the formatter without having to touch generic state.

Jim

> (But generally yes, it's better (and faster) to avoid expression evaluation altogether. I don't know how many data formatters are you dealing with, but I would recommend changing them to not do that.)
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/llvm/llvm-project/pull/129092#issuecomment-2739612203>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADUPVW3GXGYL53DSBSA5ZET2VJ6CXAVCNFSM6AAAAABYAO4MDSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMZZGYYTEMRQGM>.
> You are receiving this because you are on a team that was mentioned.
>

real-mikhail commented:
Thanks folks. I agree that approach with the option is the best (will see what looks better, something like `target.process.expressions-flush-memory-cache` or flag in `SBExpressionOptions`). Since this flag will affect very low-level code maybe global flag (in `target.process`) will be easier to pass.
And yes, this flag will be disabled by default.

Commit:
[lldb] Fixed test on Linux

Commit:
[lldb] Make new behaviour opt-in via "target.process.process-state-tracks-memory-cache" key

Commit:
[lldb] Added ability to dump ProcessModID and added test to verify that simple expressions do not cause memory_id to be bumped

Commit:
[lldb] Fixed test_all_settings_exist

Commit:
[lldb] Give new setting a better name

Commit:
[lldb] Fixed test on Linux

Commit:
[lldb] Fixed formatting

Commit:
[lldb] Give a new dump command a clearer name

real-mikhail commented:
Hi,

I added the new option key `target.process.track-memory-cache-changes` (which defaults to `true`, meaning bulletproof behaviour). Setting it to `false` enables the optimization I initially suggested. I also added a new dump command and ensured with test that MemoryID is not changed when executing simple expressions.

jimingham commented on lldb/source/Target/TargetProperties.td:
```diff
@@ -296,6 +296,9 @@ let Definition = "process" in {
     Desc<"Debugger's behavior upon fork or vfork.">;
+  def TrackMemoryCacheChanges: Property<"track-memory-cache-changes", "Boolean">,
+    DefaultTrue,
+    Desc<"If true, memory cache modifications (which happen often during expressions evaluation) will bump process state ID (and invalidate all synthetic children). Disabling this option helps to avoid synthetic children reevaluation when pretty printers heavily use expressions, but convenience variables won't reevaluate synthetic children automatically.">;
```
It's a little unclear that the "but" in the last sentence is is the downside to the "helps".   So instead of "but convenience..." you did ".  The downside of this setting is"

jimingham commented:
This looks pretty good to me.  I had one trivial doc comment, and I think it's better not to start sprinkling internal debugging commands into the lldb command set.  The debugger has enough commands already, I like to avoid adding ones that nobody but us will ever use.
If we start needing to do this a lot, then we should probably follow gdb and add a "maintenance" command to gather them all up.  That way normal user would just have to overlook one top-level command.  But in this case, we already have `process status` so the debugging dump would fit naturally as an option to that command.

Commit:
[lldb] Code review follow up

real-mikhail commented on lldb/source/Commands/CommandObjectProcess.cpp:
```diff
@@ -1827,6 +1827,33 @@ class CommandObjectMultiwordProcessTrace : public CommandObjectMultiword {
+class CommandObjectProcessDumpModificationId : public CommandObjectParsed {
+public:
+  CommandObjectProcessDumpModificationId(CommandInterpreter &interpreter)
+      : CommandObjectParsed(interpreter, "process dump-modification-id",
```
Yeah, I was reluctant to add new dump command too, but didn't think about extending other command. Fixed it (and other comment about command description too).

jimingham reviewed (approved):
LGTM, thanks for working on this!

real-mikhail commented:
Hi @JDevlieghere 

May I ask you to review the changes as well? That's my first PR in LLVM but I guess I need approvals from all reviewers before merge.

DavidSpickett commented:
Jonas is added as he's listed as the code owner for this area, but does not necessarily review all PRs.

In general, you only need one approval - https://llvm.org/docs/CodeReview.html#lgtm-how-a-patch-is-accepted

Unless you happen to know or suspect others would want to comment, but this is easy to get wrong so don't worry too much about that. Fine to ask, folks can always say they have no opinion.

I assume you don't have the ability to merge this PR yourself? I expect not given it's your first PR here.

Since you already asked @JDevlieghere, they might as well glance at it but I expect Jim's review is enough and Jonas can just merge this.

real-mikhail commented:
I assume that there will be no further comments here.

Since this is my first change in LLVM I cannot merge it myself. @jimingham may I kindly ask you to merge this PR?

jimingham commented:
Done.  Thanks for working on this.

kazutakahirata commented:
@real-mikhail @jimingham I've landed 2bff80f25d51e24d3c552e033a2863dd36ef648b to fix a warning from this PR.  Thanks!

DavidSpickett commented:
I've [reverted](https://github.com/llvm/llvm-project/commit/daa4061d61216456baa83ab404e096200e327fb4) this because it's failing on Windows on Arm.

It looks like all the IDs are there but not in the right place, so if we're lucky it's just some extra line that only appears on Windows on Arm. I'm looking into it now.

DavidSpickett commented:
It was an [ABI](https://github.com/llvm/llvm-project/commit/7d01b85c2a0aa8bc91f731cfeb2a7b6885b8d04a) difference between X86 and AArch64.

If you can think of a way to force a memory write without relying on the return address, it would be nice to get this test running on AArch64 again.

real-mikhail commented:
With the current test fix (`TestProcessModificationIdOnExpr`), it is locked to `target-x86` but it seems that it should require `target-x86_64` instead. Currently it is not run on Windows-x64 and it fails on Windows-x86 (since the test is checking low level internal data, and it is different on different architectures).
Ideally, I should not check the exact values of `m_stop_id` and `m_memory_id`. We are only interested in the fact whether those values are changed between several calls `process status -d`. But as far as I know there is no way in testing framework to "remember" specific value and refer to it (compare it) later.

DavidSpickett commented:
Please correct the REQUIRES if you have access to such machines to confirm it. Maybe the test is so specific that it is fine to be on only one architecture/os combination.

I recall a way in codegen tests to capture a line number in one pattern and re-use that in a check pattern, I'll have a look for it tomorrow.

DavidSpickett commented:
Random example - `llvm/test/Analysis/BasicAA/separate_storage.ll`:
```
define i8 @simple_no(ptr %p1, ptr %p2) {
; CHECK-LABEL: @simple_no(
; CHECK-NEXT:  entry:
; CHECK-NEXT:    store i8 0, ptr [[P1:%.*]], align 1
; CHECK-NEXT:    store i8 1, ptr [[P2:%.*]], align 1
; CHECK-NEXT:    [[LOADOFSTORE:%.*]] = load i8, ptr [[P1]], align 1
; CHECK-NEXT:    ret i8 [[LOADOFSTORE]]
;
```
https://llvm.org/docs/CommandGuide/FileCheck.html#filecheck-string-substitution-blocks

real-mikhail commented:
Thanks for the example and the link.
I fixed test to work on both Windows-x86 and Windows-x64 and created new PR: https://github.com/llvm/llvm-project/pull/138941
